### DISCUSSION TITLE
Deploying Dagster on AWS ECS with AWS Copilot

### DISCUSSION THREAD

**Discussion by ammerzon** [dagster-io/dagster#14946] [Q&A] [answered]
Hey Dagster community!

There is already a tutorial on how to deploy Dagster on AWS ECS: [Deploying in ECS](https://docs.dagster.io/deployment/guides/aws#deploying-in-ecs) with the corresponding [example project](https://github.com/dagster-io/dagster/tree/1.3.11/examples/deploy_ecs). Does anyone have experience deploying it this way using [AWS Copilot](https://aws.github.io/copilot-cli/)? Respectively, do you happen to know if this is possible at all?

**Comment by ammerzon** [comment, marked as answer]
I have now deployed Dagster with AWS Copilot, so it is possible. If anyone needs help, please ping me.

**Reply by ilkkapeltola** [reply]
I'm trying to achieve this now. Running into issues with the docker compose approach and as it has been retired, looking for the AWS Copilot solution instead.

How does your configuration look like? You have a webserved+daemon+user_code setup like the tutorial?
Are you using a single Dockerfile, or did you split these out?
I think the main concern is the manifest file, any pitfalls there I should be aware of, or will it be pretty obvious?

**Reply by ammerzon** [reply]
Hi @ilkkapeltola. The folder structure looks like this:

```
copilot/
├── daemon
│   ├── addons
│   │   ├── daemon-iam.yml
│   │   ├── dagster-assets-bucket-access-policy.yml
│   │   └── dagster-logs-bucket-access-policy.yml
│   ├── manifest.yml
│   └── overrides
│       └── cfn.patches.yml
├── environments
│   ├── addons
│   │   ├── addons.parameters.yml
│   │   ├── dagster-assets-bucket.yml
│   │   ├── dagster-database.yml
│   │   └── dagster-logs-bucket.yml
│   └── prod
│       └── manifest.yml
├── secrets.yml
├── user-code
│   ├── addons
│   │   └── dagster-logs-bucket-access-policy.yml
│   └── manifest.yml
└── webserver
    ├── addons
    │   ├── dagster-logs-bucket-access-policy.yml
    │   └── webserver-iam.yml
    ├── manifest.yml
    └── overrides
        └── cfn.patches.yml
src/
├── Dockerfile
```

So yes, I have three different components, as mentioned in the tutorial, and a single Dockerfile.
You can use this as a starting point:
[dagster-copilot.zip](https://github.com/dagster-io/dagster/files/14344946/dagster-copilot.zip)

**Reply by ilkkapeltola** [reply]
Ooooh, you're a life-saver! This saves me a ton of time. Thank you so much!

**Reply by ammerzon** [reply]
@tacastillo if you want to add this to the Dagster docs or examples. Feel free to use it 😁

**Reply by SSchneider22** [reply]
@ammerzon 
I am a beginner in AWS Copilot, I am having trouble deploying Dagster to ECS using docker compose and I want to deploy Dagster to ECS with AWS Copilot.

I have two questions.

I would like to know the code in the Dockerfile.
What commands should I actually execute using AWS Copilot CLI?

**Reply by ammerzon** [reply]
To deploy the three services use the following commands:

```sh
# Init (only needed once)
copilot app init dagster
copilot svc init -n user-code
copilot svc init -n daemon
copilot svc init -n webserver
copilot env init -n prod
# Deploy
copilot env deploy --name prod
copilot deploy -e prod -n user-code
copilot deploy -e prod -n daemon
copilot deploy -e prod -n webserver
```

Regarding the Dockerfile I used a combination of the [Poetry Dockerfile](https://github.com/orgs/python-poetry/discussions/1879#discussioncomment-7284113) and the [ECS deploy Dockerfile](https://github.com/dagster-io/dagster/blob/1.7.7/examples/deploy_ecs/Dockerfile).

**Reply by SSchneider22** [reply]
Thank you for your answer!

I will try it with what you answered.

**Reply by Jhakir-sal** [reply]
Can you share your Dockerfile?

**Reply by ammerzon** [reply]
@Jhakir-sal Unfortunately, my Dockerfile is quite customized and includes several components that likely aren't relevant to your use case. However, the base I started from was the following file: https://github.com/dagster-io/dagster/blob/79ae041145d3e994e92e2c496073acab9956569a/examples/deploy_ecs/Dockerfile

**Comment by Suman5x** [comment]
Hello @ammerzon can you help me with the copilot tutorial?

**Reply by ammerzon** [reply]
Have you tried the template files I provided? 😁

**Reply by Suman5x** [reply]
Nope i have tried the docs and the docker compose push call is still running from an hour.

**Reply by ammerzon** [reply]
I haven't tried the Docker compose approach, so I can't help you with that, but you can use the provided ZIP if you want.

**Comment by RobertCastagna** [comment]
Hi @ammerzon, I'm using your templated structure and have (tried to write) written my own Dockerfile for it. I'm getting an error:

 ```X execute Load Balanced Web Service init: check if docker engine is running: unmarshal docker info message: invalid character 'i' in literal false (expecting 'l')```
 
after running :

```.\copilot init --app app-name --name dagster --type "Load Balanced Web Service" --dockerfile "./src/Dockerfile" --port XX --deploy``` (ref:  https://docs.aws.amazon.com/AmazonECS/latest/developerguide/getting-started-aws-copilot-cli.html)

 I'm running this command from the root of the project (inside directory that contains the folders src and dagster-copilot). Any idea if this is an issue with my Dockerfile or Docker application? 

Docker version 25.0.3
aws-cli/2.0.21
Windows 10 

edit: docker engine is running, tested with ```docker ps``` showing other actively running containers

**Reply by ammerzon** [reply]
This is a bug in the AWS Copilot CLI. Here you can find a workaround: https://github.com/aws/copilot-cli/issues/5656#issuecomment-1930791131
